doc/manual/syntax.md

```diff
@@ -0,0 +1,313 @@
+"This one"
+
+> if "forty-two" == 42 then "equal?" else "unequal"
+"unequal"
+```
+
+### Let-In
+Let-in allows the binding of an expression. It is used as `let <ident> = <expr> in <expr>`.
+
+Examples:
+```
+> let r = { a = "a", b = "b" } in r.a
+"a"
+
+> let inner = { inside = true } in let outer = { outside = inner.inside } in outer.outside
+true
+
+> let a = 1 in let b = 2 in a + b
+3
+```
+
+## Functions
+A function is declared using the `fun` keyword, then arguments separated by spaces, and finally an arrow `=>` to add the body of the function.
+To call a function, just add the arguments after it separated with spaces.
+Functions in Nickel are curried, meaning that a function taking multiple arguments is actually a function that takes a single argument and returns a function taking the rest of the arguments, until it is applied.
+
+Example
+```
+> (fun a b => a + b) 1 2
+3
+
+let add = fun a b => a + b in add 1 2
+3
+
+> let add = fun a b => a + b in
+    let add1 = add 1 in
+        add1 2
+3
+```
+
+## Typing
+To give a type to a value, we write it with `< value > : < type >`.
+More information on typing in the relevant document.
+
+Examples:
+```
+5 : Num
+"Hello" : Str
+
+(fun a b => a + b) : Num -> Num -> Num
+let add : Num -> Num -> Num = fun a b => a + b
+
+{a: Num = 1, b: Bool = true, c : List Num = [ 1 ]}
+let r : {a : Num, b : Bool, c : List Num} = { a = 1, b = true, c = [ 1 ] }
+
+{ a = 1, b = 2 } : { _ : Num }
+let r : { _ : Num } = { a = 1, b = 2 }
+```
+
+## Metadata
+Metadata can be used to add contracts (more information in relevant documentation), documentation or default values.
+It is introduced with a `|`. Multiple metadata can be chained.
+
+For adding a regular type as contract, the syntax is `| < type >`. For a defined contract, we use the `| #< defined type >`.
+
+Examples:
+```
+> 5 | Num
+5
+
+> 5 | Bool
+error: Blame error: contract broken by a value.
+
+> let SmallNum = contracts.from_predicate (fun x => x < 5) in 1 | #SmallNum
+1
+
+> let SmallNum = contracts.from_predicate (fun x => x < 5) in 10 | #SmallNum
+error: Blame error: contract broken by a value.
+
+> let SmallNum = contracts.from_predicate (fun x => x < 5) in
+    let NotTooSmallNum = contracts.from_predicate (fun x => x >= 2) in
+        3 | Num | #SmallNum | #NotTooSmallNum
+3
+```
+
+Adding documentation can be done with `| doc < a string >`.  
+Examples:
+```
+> 5 | doc "The number five"
+5
+
+> true | Bool | doc m#"
+    If something is true,
+    it is based on facts rather than being invented or imagined,
+    and is accurate and reliable.
+    (Collins dictionary)
+    "#m
+true
+```
+
+Default values are used for records contracts, to give a default values to fields.
+It is noted as `| default = < default value >`.  
+Examples:
+```
+> let Ais2ByDefault = { a | default = 2 } in {} | #Ais2ByDefault
+{ a = 2 }
+
+> let Ais2ByDefault = { a | default = 2 } in { a = 1 } | #Ais2ByDefault
+{ a = 1 }
+```
```

Comment:
Could add an example of the usage of default values for merging/overriding:
```
> {foo | default = 1, bar = foo + 1} & {foo = 2}
{foo = 2, bar = 3}

Reply:
@yannham I thought about this, but I have the impression this is out of scope for the syntax document.

Reply:
Mmh. What about just putting the example, but without explaining it further? The idea is be to reflect how `default` can be encountered in the wild, but deferring any deep explanation to the relevant manual section?